File: bench/a02_indirect_call.c
/* a02_indirect_call.c — 间接调用/函数指针密集，N=100000000 次
 * 与 bench/a02_indirect_call.x / .zig 三语言同算法。
 * 4 个函数 op_add/op_sub/op_mul/op_xor 放入函数指针表，
 * 循环 N 次按 i&3 选择函数，累加结果。
 * 防常量折叠：__asm__ volatile + 返回累加值。 */
#include <stdint.h>

static int32_t op_add(int32_t a, int32_t b) { return a + b; }
static int32_t op_sub(int32_t a, int32_t b) { return a - b; }
static int32_t op_mul(int32_t a, int32_t b) { return a * b; }
static int32_t op_xor(int32_t a, int32_t b) { return a ^ b; }

int main(void) {
  int32_t (*ops[4])(int32_t, int32_t) = { op_add, op_sub, op_mul, op_xor };
  int32_t n = 100000000;
  int32_t acc = 0;
  int32_t i = 0;
  while (i < n) {
    acc = ops[i & 3](acc, i);
    i = i + 1;
    /* B-CMP：阻止 gcc/clang -O2 把间接调用去虚拟化为直调或常量折叠。 */
    __asm__ volatile("" : "+r"(i), "+r"(acc) : : "memory");
  }
  return (int)(acc & 0xFFFF);
}

File: bench/a02_indirect_call.zig
// a02_indirect_call.zig — 与 bench/a02_indirect_call.x / .c 等价的 Zig 参照
// 间接调用/函数指针密集，N=100000000 次。
// 4 个函数 op_add/op_sub/op_mul/op_xor 放入函数指针表，
// 循环 N 次按 i&3 选择函数，累加结果。
// 使用 wrapping 算术（+% / -% / *%）以匹配 C/xlang 的 i32 溢出回绕语义。
const std = @import("std");

fn op_add(a: i32, b: i32) i32 { return a +% b; }
fn op_sub(a: i32, b: i32) i32 { return a -% b; }
fn op_mul(a: i32, b: i32) i32 { return a *% b; }
fn op_xor(a: i32, b: i32) i32 { return a ^ b; }

pub fn main() !void {
    const OpFn = *const fn (i32, i32) i32;
    const ops: [4]OpFn = .{ &op_add, &op_sub, &op_mul, &op_xor };
    const n: i32 = 100_000_000;
    var acc: i32 = 0;
    var i: i32 = 0;
    while (i < n) : (i += 1) {
        acc = ops[@intCast(i & 3)](acc, i);
    }
    std.process.exit(@intCast(acc & 0xFF));
}
